File: VB/B143239/ExplorerUserControl.vb
Imports Microsoft.VisualBasic
Imports DevExpress.XtraEditors
Imports DevExpress.XtraBars
Imports System.IO
Imports System
Imports DevExpress.XtraTreeList
Imports System.Drawing

Namespace B143239
	Partial Public Class ExplorerUserControl
		Inherits XtraUserControl
		Private loadDrives As Boolean
		Private Sub New()
			InitializeComponent()
		End Sub

		Public Sub New(ByVal manager As BarManager)
			Me.New()
			CreateMenu(manager)
			treeList.DataSource = New Object()
		End Sub

		Private Sub CreateMenu(ByVal manager As BarManager)
			dockControl.BeginUpdate()
			manager.DockControls.Add(dockControl)

			Dim bar As New Bar(manager, "explorerBar")
			bar.BeginUpdate()
			bar.StandaloneBarDockControl = dockControl
			bar.DockStyle = BarDockStyle.Standalone
			bar.CanDockStyle = BarCanDockStyle.Standalone
			bar.OptionsBar.DisableCustomization = True

			Dim item As New BarButtonItem(manager, "View properties")
			AddHandler item.ItemClick, AddressOf OnViewPropertiesItemClick
			bar.ItemLinks.Add(item)

			bar.EndUpdate()
			dockControl.EndUpdate()
		End Sub

		Private Sub OnViewPropertiesItemClick(ByVal sender As Object, ByVal e As ItemClickEventArgs)
			If treeList.FocusedNode Is Nothing Then
				Return
			End If
			RaiseNeedDisplayProperty(New DirectoryInfo(CStr(treeList.FocusedNode(colFullName))))
		End Sub

		Private Shared fNeedDisplayProperty As Object = New Object()
		Public Custom Event NeedDisplayProperty As NeedDisplayPropertyEventHandler
			AddHandler(ByVal value As NeedDisplayPropertyEventHandler)
				Events.AddHandler(fNeedDisplayProperty, value)
			End AddHandler
			RemoveHandler(ByVal value As NeedDisplayPropertyEventHandler)
				Events.RemoveHandler(fNeedDisplayProperty, value)
			End RemoveHandler
			RaiseEvent(ByVal sender As Object, ByVal e As NeedDisplayPropertyEventArgs)
			End RaiseEvent
		End Event
		Public Sub RaiseNeedDisplayProperty(ByVal directory As DirectoryInfo)
			Dim handler As NeedDisplayPropertyEventHandler = TryCast(Events(fNeedDisplayProperty), NeedDisplayPropertyEventHandler)
			If handler IsNot Nothing Then
				handler(Me, New NeedDisplayPropertyEventArgs(directory))
			End If
		End Sub

		Private Function IsFile(ByVal info As DirectoryInfo) As Boolean
			Return (info.Attributes And FileAttributes.Directory) = 0
		End Function

		Private Sub OnTreeListVirtualTreeGetCellValue(ByVal sender As Object, ByVal e As VirtualTreeGetCellValueInfo) Handles treeList.VirtualTreeGetCellValue
			Dim di As New DirectoryInfo(CStr(e.Node))
			If e.Column Is colName Then
				e.CellData = di.Name
			End If
			If e.Column Is colType Then
				If (Not IsFile(di)) Then
					e.CellData = "Folder"
				Else
					e.CellData = "File"
				End If
			End If
			If e.Column Is colFullName Then
				e.CellData = di.FullName
			End If
		End Sub

		Private Sub OnTreeListVirtualTreeGetChildNodes(ByVal sender As Object, ByVal e As VirtualTreeGetChildNodesInfo) Handles treeList.VirtualTreeGetChildNodes
			If (Not loadDrives) Then ' create drives
				Dim root() As String = Directory.GetLogicalDrives()
				e.Children = root
				loadDrives = True
			Else
				Try
					Dim path As String = CStr(e.Node)
					If Directory.Exists(path) Then
						Dim dirs() As String = Directory.GetDirectories(path)
						Dim files() As String = Directory.GetFiles(path)
						Dim arr(dirs.Length + files.Length - 1) As String
						dirs.CopyTo(arr, 0)
						files.CopyTo(arr, dirs.Length)
						e.Children = arr
					Else
						e.Children = New Object() { }
					End If
				Catch
					e.Children = New Object() { }
				End Try
			End If
		End Sub
	End Class

	Public Delegate Sub NeedDisplayPropertyEventHandler(ByVal sender As Object, ByVal e As NeedDisplayPropertyEventArgs)
	Public Class NeedDisplayPropertyEventArgs
		Inherits EventArgs
		Public Sub New(ByVal directory As DirectoryInfo)
			fDirectory = directory
		End Sub
		Private fDirectory As DirectoryInfo
		Public ReadOnly Property Directory() As DirectoryInfo
			Get
				Return fDirectory
			End Get
		End Property
	End Class
End Namespace

File: CS/B143239/ExplorerUserControl.cs
using DevExpress.XtraEditors;
using DevExpress.XtraBars;
using System.IO;
using System;
using DevExpress.XtraTreeList;
using System.Drawing;

namespace B143239 {
    public partial class ExplorerUserControl :XtraUserControl {
        private bool loadDrives;
        private ExplorerUserControl() {
            InitializeComponent();
        }

        public ExplorerUserControl(BarManager manager)
            : this() {
            CreateMenu(manager);
            treeList.DataSource = new object();
        }

        private void CreateMenu(BarManager manager) {
            dockControl.BeginUpdate();
            manager.DockControls.Add(dockControl);

            Bar bar = new Bar(manager, "explorerBar");
            bar.BeginUpdate();
            bar.StandaloneBarDockControl = dockControl;
            bar.DockStyle = BarDockStyle.Standalone;
            bar.CanDockStyle = BarCanDockStyle.Standalone;
            bar.OptionsBar.DisableCustomization = true;

            BarButtonItem item = new BarButtonItem(manager, "View properties");
            item.ItemClick += new ItemClickEventHandler(OnViewPropertiesItemClick);
            bar.ItemLinks.Add(item);

            bar.EndUpdate();
            dockControl.EndUpdate();
        }

        private void OnViewPropertiesItemClick(object sender, ItemClickEventArgs e) {
            if (treeList.FocusedNode == null) return;
            RaiseNeedDisplayProperty(new DirectoryInfo((string)treeList.FocusedNode[colFullName]));
        }

        private static object fNeedDisplayProperty = new object();
        public event NeedDisplayPropertyEventHandler NeedDisplayProperty {
            add { Events.AddHandler(fNeedDisplayProperty, value); }
            remove { Events.RemoveHandler(fNeedDisplayProperty, value); }
        }
        public void RaiseNeedDisplayProperty(DirectoryInfo directory) {
            NeedDisplayPropertyEventHandler handler =
                Events[fNeedDisplayProperty] as NeedDisplayPropertyEventHandler;
            if (handler != null) handler(this, new NeedDisplayPropertyEventArgs(directory));
        }

        bool IsFile(DirectoryInfo info) {
            return (info.Attributes & FileAttributes.Directory) == 0;
        }

        private void OnTreeListVirtualTreeGetCellValue(object sender, VirtualTreeGetCellValueInfo e) {
            DirectoryInfo di = new DirectoryInfo((string)e.Node);
            if (e.Column == colName)
                e.CellData = di.Name;
            if (e.Column == colType) {
                if (!IsFile(di))
                    e.CellData = "Folder";
                else
                    e.CellData = "File";
            } if (e.Column == colFullName)
                e.CellData = di.FullName;
        }

        private void OnTreeListVirtualTreeGetChildNodes(object sender, VirtualTreeGetChildNodesInfo e) {
            if (!loadDrives) { // create drives
                string[] root = Directory.GetLogicalDrives();
                e.Children = root;
                loadDrives = true;
            } else {
                try {
                    string path = (string)e.Node;
                    if (Directory.Exists(path)) {
                        string[] dirs = Directory.GetDirectories(path);
                        string[] files = Directory.GetFiles(path);
                        string[] arr = new string[dirs.Length + files.Length];
                        dirs.CopyTo(arr, 0);
                        files.CopyTo(arr, dirs.Length);
                        e.Children = arr;
                    } else e.Children = new object[] { };
                } catch { e.Children = new object[] { }; }
            }
        }
    }

    public delegate void NeedDisplayPropertyEventHandler(object sender, NeedDisplayPropertyEventArgs e);
    public class NeedDisplayPropertyEventArgs :EventArgs {
        public NeedDisplayPropertyEventArgs(DirectoryInfo directory) {
            fDirectory = directory;
        }
        private DirectoryInfo fDirectory;
        public DirectoryInfo Directory { get { return fDirectory; } }
    }
}
